synapse/appservice/__init__.py

```diff
@@ -32,6 +35,20 @@ class ApplicationServiceState:
     UP = "up"
 
 
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class Namespace:
+    exclusive: bool
+    group_id: Optional[str]
+    regex: Pattern
+
+
+@attr.s(slots=True, frozen=True, auto_attribs=True)
+class Namespaces:
+    users: List[Namespace]
+    aliases: List[Namespace]
+    rooms: List[Namespace]
```

Comment:
This class doesn't seem to add much, I wonder if we should just inline these onto the `ApplicationService` class.

Reply:
As a data point, I could have used it when writing tests at https://github.com/matrix-org/synapse/pull/11215/files#diff-0b8c043365754e58302f0701c4e2f273dc190f516a326b42496b68a09fe26ef6R555 :slightly_smiling_face: 

But an `Dict[ApplicationService.NS_LIST, Namespace]` isn't too bad to write either.

Reply:
> Dict[ApplicationService.NS_LIST, Namespace]

Oh, I assumed this would restrict the entries to just the three types of allowed namespace strings, rather than allowing any str. But perhaps that's not possible.

Reply:
You need to use an enum to do that -- that could be a solution though.

Reply:
Oh right, yes indeed! That could be useful to have available.

Reply:
> You need to use an enum to do that -- that could be a solution though.

Or a `Literal["name1", "name2", "name3"]`.

I was really keen to use `StrEnum` but Sean found that it didn't work well with interning. See #11356.